cargo-marker: add initial configuration support

Niki4tap opened the pull request:
Tracked in #81

Adds initial configuration implementation, checking out one task :)

@xFrednet

xFrednet commented on cargo-marker/src/main.rs:
```diff
@@ -77,7 +110,16 @@ fn main() -> Result<(), ExitStatus> {
             .take_while(|s| s != CARGO_ARGS_SEPARATOR),
     );
 
+    let config = match Config::get_marker_config() {
+        Ok(v) => Some(v),
+        Err(e) => match e {
+            config::ConfigFetchError::NotFound => None,
+            _ => return Err(e.emit_and_convert()),
+        },
+    };
```
This should be moved after the flags and version check or maybe better move this into `choose_lint_crates` method, to only load it, if the user doesn't specify lints in the CLI. 

A reason to keep it in main would be, that later `install_driver` might want to also take the configuration, to allow overriding the driver version. But that's in the future and then we can load it here on demand as well.

Either solution is fine, but it should be after the version flag check

Niki4tap replied:
You're right, nice catch! Totally overlooked it for some reason.

Niki4tap commented on cargo-marker/src/config.rs:
```diff
@@ -0,0 +1,147 @@
+
+impl Config {
+    pub fn get_marker_config() -> Result<Config, ConfigFetchError> {
+        let Ok(mut config_file) = File::open(CARGO_TOML) else {
```
Oh, didn't know about cargo metadata, but I think invoking cargo would already be slower, and metadata contains a lot more than just the config, so I think parsing it ourselves is the right choice here

Niki4tap commented on cargo-marker/src/config.rs:
```diff
@@ -0,0 +1,147 @@
+        Ok(marker_config)
+    }
+
+    pub fn collect_paths(&self) -> Result<Vec<String>, ExitStatus> {
```
It is safe, [toml spec](https://toml.io/en/v1.0.0) requires that...
> A TOML file must be a valid UTF-8 encoded Unicode document.

And we aren't doing any additional processing here, just parsing the config

Niki4tap commented on Cargo.toml:
```diff
@@ -4,6 +4,9 @@ members = [
 ]
+
+[workspace.metadata.marker.lints]
+dogfood = { path = "marker_lints" }
```
Oh, good, another thing I didn't know about cargo :sweat_smile:

It looks like `cargo` allows it, if you also have `package = "marker_lints"`, which I have apparently missed here and in the implementation (whoops), I'll add the missing fields and I think I'll enforce the dependency renaming in the way `cargo` does, i.e. error if the package name isn't the same as renamed dependency, or the `package` field, if it exists.

xFrednet replied:
Renaming lint crates was discussed in the design repo, I'd like to have the option, but generally enforce it like cargo does. We also need to know the binary name to pass it to the adapter for loading :)

xFrednet commented on cargo-marker/src/config.rs:
```diff
@@ -0,0 +1,154 @@
+            return Err(ConfigFetchError::IoError(e));
+        }
+
+        let cargo_config = match from_str::<Value>(&config_str) {
```
Just a general comment to think about for the future:

We could split this function, to only the file loading part and the thing that parses the text. The advantage would be testability. So far, there are no real tests for `cargo-marker` and for development it's good enough if it just works. So, you don't have to add them as part of this PR. It's just something we should think about later down the line :)

Niki4tap replied:
Done

xFrednet commented on cargo-marker/src/lints.rs:
```diff
@@ -6,22 +6,51 @@ use std::{
 
 use crate::ExitStatus;
 
+pub struct LintCrateSpec<'a> {
```
Using an extra struct is more readable than the alternative of using a tuple. It should also make it easier to add crates via git etc. Could you add doc comments, explaining why and when package name can be `None`? Also, mentioning that we require the package name to be UTF-8 could be useful. TOML, forces this, but still worth a mention. That will make later audits a bit easier. (If that ever happens :))

Niki4tap replied:
> Also, mentioning that we require the package name to be UTF-8 could be useful

`str` implies that, so I feel that would be more explicit than it should be. 

> Could you add doc comments, explaining why and when package name can be `None`?

Sure, I'll add some doc comments

xFrednet replied:
> str implies that, so I feel that would be more explicit than it should be.

Well, if someone else reads it, it could be good to know that this was an intentional choice. Using stings for paths and file names, is quite common even if it's not always correct. There is #67 because I didn't think about that during the first implementation. That's also why I noticed it.

xFrednet commented on cargo-marker/src/lints.rs:
```diff
@@ -6,22 +6,51 @@ use std::{
-pub fn build_local_lint_crate(crate_dir: &Path, target_dir: &Path, verbose: bool) -> Result<PathBuf, ExitStatus> {
-    if !crate_dir.exists() {
-        eprintln!("The given lint can't be found, searched at: `{}`", crate_dir.display());
+pub fn build_local_lint_crate(
```
Does this still need to be public?

Niki4tap replied:
Nope

Niki4tap commented:
Thanks again for the review!

I addressed the comments, did some refactoring, and I'm relatively happy with the code now, can you review it again please?

xFrednet commented on cargo-marker/src/lints.rs:
```diff
@@ -6,22 +6,51 @@ use std::{
+    }
+
+    /// Currently only checks for semicolons, can be extended in the future
+    pub fn validate(&self) -> bool {
```
I think the polarity of the return is confusing. I was surprised, that a later check `crate_entries.iter().any(LintCrateSpec::validate)` errored if any call returns true. I would have expected `true` to indicate that everything has been validated. Maybe `is_valid` would be a better name.

Niki4tap replied:
Oh, yeah, you're right, that totally needs to be renamed :sweat_smile:

xFrednet reviewed (commented):
Generally, looks good to me, a few more NITs (Sorry if I'm too picky) and then this should be ready to go :)

xFrednet reviewed (approved):
Looks good to me. Thank you!!

bors r+
